### PR TITLE
Color engagement table values based on categories.

### PR DESCRIPTION
**Pull request by dsjen**
table

@dan-f please review

### CODE REVIEW

File: analytics_dashboard/static/apps/learners/js/spec/utils-spec.js [hunk 1]
```diff
@@ -55,5 +55,28 @@ require([
                 jasmine.clock().tick(2);
             });
         });
+
+        describe('inRange', function () {
+
+            it('return true when value is in range', function () {
```
Comment (dan-f) [inline analytics_dashboard/static/apps/learners/js/spec/utils-spec.js]: Nit: "return" -> "returns"

File: analytics_dashboard/static/apps/learners/js/spec/utils-spec.js [hunk 2]
```diff
@@ -55,5 +55,28 @@ require([
...
+                expect(LearnerUtils.inRange(1, [0, 2])).toBe(true);
+                expect(LearnerUtils.inRange(10.5, [5, 11])).toBe(true);
+                expect(LearnerUtils.inRange(44, [null, 45])).toBe(true);
+                expect(LearnerUtils.inRange(59.3, [59, null])).toBe(true);
+            });
+
+            it('return true when value is out of range', function () {
```
Comment (dan-f) [inline analytics_dashboard/static/apps/learners/js/spec/utils-spec.js]: I think you mean "returns false" here :)

File: analytics_dashboard/static/apps/learners/js/utils.js [hunk 3]
```diff
@@ -21,6 +21,32 @@ define([], function () {
...
+            var min = range[0],
+                max = range[1],
+                isRangeValid = !_(min).isNull() || !_(max).isNull(),
+                isFound = false;
+
+            // TODO: if null min is unbounded and null max is unbounded, then does this become true?!?!
+            // problem is that we're returning null for both values currently
+            // should we explicitly say that if they're both null then we don't evaluate?
+
+            if (isRangeValid && !_(value).isNull()) {
```
Comment (dan-f) [inline analytics_dashboard/static/apps/learners/js/utils.js]: Why do we need to test if `value` is null?

File: analytics_dashboard/static/apps/learners/js/utils.js [hunk 4]
```diff
@@ -21,6 +21,32 @@ define([], function () {
...
+                isFound = _(min).isNull() ? true : value >= min;
```
Comment (dan-f) [inline analytics_dashboard/static/apps/learners/js/utils.js]: Nit: you might be able to refactor this a bit by saving the value of `_(min).isNull()` into a variable with a semantic name such as `minIsUnbounded` or something along those lines. Same with `max`.

Comment (dan-f) [inline analytics_dashboard/static/apps/learners/js/utils.js]: Sorry to be such a nitpicker here, but I'm finding the logic of this method a bit difficult to follow, though it is quite terse, which I appreciate. Here's how I think I was thinking of it:

``` javascript
var minIsUnbounded = _.isNull(min),
    maxIsUnbounded = _.isNull(max);

if (minIsUnbounded && maxIsUnbounded) {
    throw new Error('min and max range values cannot both be null (unbounded)');
}

if (minIsUnbounded) {
    return value < max;
} else if (maxIsUnbounded) {
    return value >= min;
} else {
    return value >= min && value < max;
}
```

Feel free to take this or leave it!

File: analytics_dashboard/static/apps/learners/js/views/roster-view.js [hunk 5]
```diff
@@ -141,15 +141,34 @@ define([
      * Factory for creating a Backgrid cell class that renders a key
      * from the learner model's engagement attribute.
      */
-    createEngagementCell = function (key) {
+    createEngagementCell = function (key, options) {
         return Backgrid.Cell.extend({
+
             className: 'learner-engagement-cell ' + key,
```
Comment (dan-f) [inline analytics_dashboard/static/apps/learners/js/views/roster-view.js:147]: Another way to implement this would be to encapsulate the logic in `className`, which means you wouldn't have to manually call `this.$el.addClass` in `render`. Take it or leave it :)

**Comment by dan-f** [comment]
Looking good! Could you add a behavioral test to `roster-view-spec.js`? Something where we render the table and expect that certain cells have certain classes?

File: analytics_dashboard/static/apps/learners/js/spec/utils-spec.js (hunk 1 above)
Comment (dsjen) [inline analytics_dashboard/static/apps/learners/js/spec/utils-spec.js]: Thanks!

File: analytics_dashboard/static/apps/learners/js/spec/utils-spec.js (hunk 2 above)
Comment (dsjen) [inline analytics_dashboard/static/apps/learners/js/spec/utils-spec.js]: :)

File: analytics_dashboard/static/apps/learners/js/utils.js (hunk 4 above)
Comment (dsjen) [inline analytics_dashboard/static/apps/learners/js/utils.js]: That seems quite a bit cleaner.  I'll take it!

**Comment by dsjen** [comment]
Thanks, I'm adding tests to `roster-view-spec.js`.

**Comment by dsjen** [comment]
@dan-f -- I just pushed an update that address your comments.  Adding a tests to `roster-view-spec.js` was helpful in finding a spelling error!

File: analytics_dashboard/static/apps/learners/js/spec/roster-view-spec.js [hunk 6]
```diff
@@ -140,6 +140,66 @@ define([
...
+                .toHaveClass('learner-cell-below-average');
+            expect($tr.find('td.problems_completed'))
+                .toHaveClass('learner-cell-average');
+            expect($tr.find('td.problems_attempted'))
+                .toHaveClass('learner-cell-above-average');
+            expect($tr.find('td.problem_attempts_per_completed'))
+                .toHaveClass('learner-cell-below-average');
+            expect($tr.find('td.videos_viewed'))
+                .toHaveClass('learner-cell-average');
+        });
```
Comment (dan-f) [inline analytics_dashboard/static/apps/learners/js/spec/roster-view-spec.js:201]: :clap: !! thank you this is beautiful!

**Comment by dan-f** [comment]
Great!! :+1: thanks for adding that test case!
